File: 2013 University of Chicago Invitational Programming Contest/goatropes.cs
using System;
using System.Collections.Generic;

public struct Point {
    public double x, y;
    
    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }
}

public class LinearProgramResult {
    public double Value { get; set; }
    public double[] Solution { get; set; }
}

public class Program {
    static double EuclideanDist(Point a, Point b) {
        return Math.Sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y));
    }
    
    static LinearProgramResult LinearProgramSolution(double[][] A, double[] b, double[] c) {
        int m = b.Length;
        int n = c.Length;
        
        double[][] tableau = new double[m + 2][];
        for (int i = 0; i < m + 2; i++) {
            tableau[i] = new double[n + 2];
        }
        
        int[] basic = new int[m];
        int[] nonBasic = new int[n + 1];
        
        // Initialize tableau
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                tableau[i][j] = A[i][j];
            }
            basic[i] = n + i;
            tableau[i][n] = -1;
            tableau[i][n + 1] = b[i];
        }
        
        for (int j = 0; j <= n; j++) {
            nonBasic[j] = j;
        }
        nonBasic[n] = -1;
        
        for (int j = 0; j < n; j++) {
            tableau[m][j] = -c[j];
        }
        tableau[m + 1][n] = 1;
        
        void Pivot(int row, int col) {
            double inv = 1.0 / tableau[row][col];
            
            for (int i = 0; i < m + 2; i++) {
                if (i != row && Math.Abs(tableau[i][col]) > 1e-8) {
                    double temp = tableau[i][col] * inv;
                    for (int j = 0; j < n + 2; j++) {
                        tableau[i][j] -= tableau[row][j] * temp;
                    }
                    tableau[i][col] = tableau[row][col] * temp;
                }
            }
            
            for (int i = 0; i < m + 2; i++) {
                if (i != row) {
                    tableau[i][col] *= -inv;
                }
            }
            
            for (int j = 0; j < n + 2; j++) {
                if (j != col) {
                    tableau[row][j] *= inv;
                }
            }
            tableau[row][col] = inv;
            
            int tempVal = basic[row];
            basic[row] = nonBasic[col];
            nonBasic[col] = tempVal;
        }
        
        bool Simplex(int phase) {
            while (true) {
                int r = m + (phase == 1 ? 1 : 0);
                int col = -1;
                
                for (int j = 0; j <= n; j++) {
                    if (phase == 2 && nonBasic[j] == -1) continue;
                    if (col == -1 || tableau[r][j] < tableau[r][col] - 1e-8 || 
                        (Math.Abs(tableau[r][j] - tableau[r][col]) < 1e-8 && nonBasic[j] < nonBasic[col])) {
                        col = j;
                    }
                }
                
                if (tableau[r][col] >= -1e-8) return true;
                
                int row = -1;
                for (int i = 0; i < m; i++) {
                    if (tableau[i][col] <= 1e-8) continue;
                    if (row == -1 || 
                        tableau[i][n + 1] / tableau[i][col] < tableau[row][n + 1] / tableau[row][col] - 1e-8 ||
                        (Math.Abs(tableau[i][n + 1] / tableau[i][col] - tableau[row][n + 1] / tableau[row][col]) < 1e-8 && 
                         basic[i] < basic[row])) {
                        row = i;
                    }
                }
                
                if (row == -1) return false;
                Pivot(row, col);
            }
        }
        
        int row = 0;
        for (int i = 1; i < m; i++) {
            if (tableau[i][n + 1] < tableau[row][n + 1]) {
                row = i;
            }
        }
        
        if (tableau[row][n + 1] <= -1e-8) {
            Pivot(row, n);
            if (!Simplex(1) || tableau[m + 1][n + 1] < -1e-8) {
                return new LinearProgramResult { Value = double.NegativeInfinity };
            }
            
            for (int i = 0; i < m; i++) {
                if (basic[i] == -1) {
                    int col = -1;
                    for (int j = 0; j <= n; j++) {
                        if (col == -1 || tableau[i][j] < tableau[i][col] - 1e-8 ||
                            (Math.Abs(tableau[i][j] - tableau[i][col]) < 1e-8 && nonBasic[j] < nonBasic[col])) {
                            col = j;
                        }
                    }
                    Pivot(i, col);
                }
            }
        }
        
        if (!Simplex(2)) {
            return new LinearProgramResult { Value = double.PositiveInfinity };
        }
        
        double[] solution = new double[n];
        for (int i = 0; i < m; i++) {
            if (basic[i] < n) {
                solution[basic[i]] = tableau[i][n + 1];
            }
        }
        
        return new LinearProgramResult { 
            Value = tableau[m][n + 1],
            Solution = solution
        };
    }
    
    public static void Main() {
        int n = int.Parse(Console.ReadLine());
        Point[] coords = new Point[n];
        
        for (int i = 0; i < n; i++) {
            string[] parts = Console.ReadLine().Split();
            coords[i] = new Point(double.Parse(parts[0]), double.Parse(parts[1]));
        }
        
        int m = n * (n - 1) / 2;
        double[][] A = new double[m][];
        for (int i = 0; i < m; i++) {
            A[i] = new double[n];
        }
        
        double[] b = new double[m];
        double[] c = new double[n];
        
        for (int i = 0; i < n; i++) {
            c[i] = 1.0;
        }
        
        int k = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                A[k][i] = 1.0;
                A[k][j] = 1.0;
                b[k] = EuclideanDist(coords[i], coords[j]);
                k++;
            }
        }
        
        LinearProgramResult result = LinearProgramSolution(A, b, c);
        Console.WriteLine($"{result.Value:F2}");
    }
}

File: 2013 University of Chicago Invitational Programming Contest/goatropes.py
import math
import sys

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

def euclidean_dist(a, b):
    return math.sqrt((a.x - b.x) ** 2 + (a.y - b.y) ** 2)

def linear_program_solution(A, b, c):
    m = len(b)
    n = len(c)
    
    tableau = [[0.0] * (n + 2) for _ in range(m + 2)]
    basic = [0] * m
    non_basic = [0] * (n + 1)
    
    # Initialize tableau
    for i in range(m):
        for j in range(n):
            tableau[i][j] = A[i][j]
        basic[i] = n + i
        tableau[i][n] = -1
        tableau[i][n + 1] = b[i]
    
    for j in range(n + 1):
        non_basic[j] = j
    non_basic[n] = -1
    
    for j in range(n):
        tableau[m][j] = -c[j]
    tableau[m + 1][n] = 1
    
    def pivot(row, col):
        inv = 1.0 / tableau[row][col]
        
        for i in range(m + 2):
            if i != row and abs(tableau[i][col]) > 1e-8:
                temp = tableau[i][col] * inv
                for j in range(n + 2):
                    tableau[i][j] -= tableau[row][j] * temp
                tableau[i][col] = tableau[row][col] * temp
        
        for i in range(m + 2):
            if i != row:
                tableau[i][col] *= -inv
        
        for j in range(n + 2):
            if j != col:
                tableau[row][j] *= inv
        tableau[row][col] = inv
        
        basic[row], non_basic[col] = non_basic[col], basic[row]
    
    def simplex(phase):
        while True:
            r = m + (1 if phase == 1 else 0)
            col = -1
            
            for j in range(n + 1):
                if phase == 2 and non_basic[j] == -1:
                    continue
                if col == -1 or tableau[r][j] < tableau[r][col] - 1e-8 or (
                    abs(tableau[r][j] - tableau[r][col]) < 1e-8 and non_basic[j] < non_basic[col]):
                    col = j
            
            if tableau[r][col] >= -1e-8:
                return True
            
            row = -1
            for i in range(m):
                if tableau[i][col] <= 1e-8:
                    continue
                if row == -1 or (
                    tableau[i][n + 1] / tableau[i][col] < tableau[row][n + 1] / tableau[row][col] - 1e-8 or (
                    abs(tableau[i][n + 1] / tableau[i][col] - tableau[row][n + 1] / tableau[row][col]) < 1e-8 and 
                    basic[i] < basic[row])):
                    row = i
            
            if row == -1:
                return False
            pivot(row, col)
    
    row = 0
    for i in range(1, m):
        if tableau[i][n + 1] < tableau[row][n + 1]:
            row = i
    
    if tableau[row][n + 1] <= -1e-8:
        pivot(row, n)
        if not simplex(1) or tableau[m + 1][n + 1] < -1e-8:
            return float('-inf'), []
        
        for i in range(m):
            if basic[i] == -1:
                col = -1
                for j in range(n + 1):
                    if col == -1 or tableau[i][j] < tableau[i][col] - 1e-8 or (
                        abs(tableau[i][j] - tableau[i][col]) < 1e-8 and non_basic[j] < non_basic[col]):
                        col = j
                pivot(i, col)
    
    if not simplex(2):
        return float('inf'), []
    
    solution = [0.0] * n
    for i in range(m):
        if basic[i] < n:
            solution[basic[i]] = tableau[i][n + 1]
    
    return tableau[m][n + 1], solution

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    
    n = int(data[0])
    idx = 1
    coords = []
    for i in range(n):
        x = float(data[idx]); y = float(data[idx + 1]); idx += 2
        coords.append(Point(x, y))
    
    m = n * (n - 1) // 2
    A = [[0.0] * n for _ in range(m)]
    b = [0.0] * m
    c = [1.0] * n
    
    k = 0
    for i in range(n):
        for j in range(i + 1, n):
            A[k][i] = 1.0
            A[k][j] = 1.0
            b[k] = euclidean_dist(coords[i], coords[j])
            k += 1
    
    result, _ = linear_program_solution(A, b, c)
    print(f"{result:.2f}")

if __name__ == "__main__":
    main()